Compiler error when building global arrays of empty Structs

### Describe the bug

Consider the following code:

```
❯ cat 1.v
module main

const num_elements = 10

struct DummyStruct {}

__global (
        d [num_elements]DummyStruct
)

fn main() {}
```

When built with the following invocation, the compilation fails:

```
❯ v -enable-globals 1.v
==================
/tmp/v_501/1.01J1JR6W82J8GJBYBQHK0EWBH8.tmp.c:2068:41: error: initializer for aggregate with no elements requires explicit braces
 Array_fixed_main__DummyStruct_10  d = {0}; // global4
                                        ^
1 error generated.
...
==================
(Use `v -cg` to print the entire error message)

builder error:
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .

```

On the other hand, if DummyStruct is non-empty then the compilation succeeds. 

Consider the following:

```
❯ cat 2.v
module main

const num_elements = 10

struct DummyStruct {
        dummy_item i32
}

__global (
        d [num_elements]DummyStruct
)

fn main() {}
```

The only change is that DummyStruct now has a dummy_item.

The compilation succeeds as follows:

```
❯ ~/Work/repos/vlang/v/v -enable-globals 2.v
```

**NOTE**:

```
❯ v --version
V 0.4.6 23e3894

❯ neofetch --off
spinlock@deadlock.local
-------------------
OS: macOS 14.5 23F79 arm64
Host: Mac14,9
Kernel: 23.5.0
Uptime: 11 days, 11 hours, 48 mins
Packages: 250 (port)
Shell: fish 3.7.1
Resolution: 5120x1440 , 3024x1964
DE: Aqua
WM: Quartz Compositor
WM Theme: Blue (Dark)
Terminal: tmux
CPU: Apple M2 Pro
GPU: Apple M2 Pro
Memory: 4652MiB / 32768MiB

```

### Reproduction Steps

See above

### Expected Behavior

Structs containing no members is a perfectly valid use case and is often referenced in the official documentation. There is nothing special about them that warrants a compiler build failure in the case of global arrays.

### Current Behavior

See bug description above.

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

0.4.6 23e3894

### Environment details (OS name and version, etc.)

macOS 14.5 arm64
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/ast/types.v
+++ b/vlib/v/ast/types.v
@@ -925,6 +925,16 @@ pub fn (t &ArrayFixed) is_compatible(t2 ArrayFixed) bool {
 	return t.size == t2.size && t.elem_type == t2.elem_type
 }
 
+pub fn (t &TypeSymbol) is_empty_struct_array() bool {
+	if t.info is ArrayFixed {
+		elem_sym := global_table.final_sym(t.info.elem_type)
+		if elem_sym.info is Struct {
+			return elem_sym.info.fields.len == 0
+		}
+	}
+	return false
+}
+
 pub fn (t &TypeSymbol) is_array_fixed() bool {
 	if t.info is ArrayFixed {
 		return true

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -4291,7 +4291,7 @@ fn (mut g Gen) debugger_stmt(node ast.DebuggerStmt) {
 fn (mut g Gen) enum_decl(node ast.EnumDecl) {
 	enum_name := util.no_dots(node.name)
 	is_flag := node.is_flag
-	if g.pref.ccompiler == 'msvc' {
+	if g.is_cc_msvc {
 		mut last_value := '0'
 		enum_typ_name := g.table.get_type_name(node.typ)
 		g.enum_typedefs.writeln('')
@@ -6274,6 +6274,8 @@ fn (mut g Gen) global_decl(node ast.GlobalDecl) {
 			default_initializer := g.type_default(field.typ)
 			if default_initializer == '{0}' && should_init {
 				def_builder.write_string(' = {0}')
+			} else if default_initializer == '{EMPTY_STRUCT_INITIALIZATION}' && should_init {
+				init = '\tmemcpy(${field.name}, (${styp}){${g.type_default(field.typ)}}, sizeof(${styp})); // global'
 			} else {
 				if field.name !in ['as_cast_type_indexes', 'g_memory_block', 'global_allocator'] {
 					init = '\t${field.name} = *(${styp}*)&((${styp}[]){${g.type_default(field.typ)}}[0]); // global'
@@ -7157,7 +7159,13 @@ fn (mut g Gen) type_default(typ_ ast.Type) string {
 		.string {
 			return '(string){.str=(byteptr)"", .is_lit=1}'
 		}
-		.interface_, .sum_type, .array_fixed, .multi_return, .thread {
+		.array_fixed {
+			if g.pref.ccompiler_type == .clang && sym.is_empty_struct_array() {
+				return '{EMPTY_STRUCT_INITIALIZATION}'
+			}
+			return '{0}'
+		}
+		.interface_, .sum_type, .multi_return, .thread {
 			return '{0}'
 		}
 		.alias {

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -7160,7 +7160,7 @@ fn (mut g Gen) type_default(typ_ ast.Type) string {
 			return '(string){.str=(byteptr)"", .is_lit=1}'
 		}
 		.array_fixed {
-			if g.pref.ccompiler_type == .clang && sym.is_empty_struct_array() {
+			if sym.is_empty_struct_array() {
 				return '{EMPTY_STRUCT_INITIALIZATION}'
 			}
 			return '{0}'

--- a/vlib/v/gen/c/dumpexpr.v
+++ b/vlib/v/gen/c/dumpexpr.v
@@ -227,11 +227,16 @@ fn (mut g Gen) dump_expr_definitions() {
 		surrounder.builder_write_afters(mut dump_fns)
 		if is_fixed_arr_ret {
 			tmp_var := g.new_tmp_var()
+			init_str := if dump_sym.is_empty_struct_array() {
+				'{EMPTY_STRUCT_INITIALIZATION}'
+			} else {
+				'{0}'
+			}
 			if typ.is_ptr() {
-				dump_fns.writeln('\t${str_dumparg_ret_type} ${tmp_var} = HEAP(${g.typ(typ.set_nr_muls(0))}, {0});')
+				dump_fns.writeln('\t${str_dumparg_ret_type} ${tmp_var} = HEAP(${g.typ(typ.set_nr_muls(0))}, ${init_str});')
 				dump_fns.writeln('\tmemcpy(${tmp_var}->ret_arr, dump_arg, sizeof(${str_dumparg_type}));')
 			} else {
-				dump_fns.writeln('\t${str_dumparg_ret_type} ${tmp_var} = {0};')
+				dump_fns.writeln('\t${str_dumparg_ret_type} ${tmp_var} = ${init_str};')
 				dump_fns.writeln('\tmemcpy(${tmp_var}.ret_arr, dump_arg, sizeof(${str_dumparg_type}));')
 			}
 			dump_fns.writeln('\treturn ${tmp_var};')

--- a/vlib/v/tests/array_fixed_empty_struct_test.v
+++ b/vlib/v/tests/array_fixed_empty_struct_test.v
@@ -1,0 +1,16 @@
+@[has_globals]
+module main
+
+const num_elements = 10
+
+struct DummyStruct {
+}
+
+__global (
+	d [num_elements]DummyStruct
+)
+
+fn test_main() {
+	a := dump(d)
+	assert a.len == num_elements
+}

